src/valid/handles.rs

```diff
@@ -0,0 +1,631 @@
+//! Implementation of [`super::Validator::validate_module_handles`].
+
+use crate::{arena::BadHandle, Handle, Range};
+
+#[cfg(feature = "validate")]
+use crate::{Arena, UniqueArena};
+
+#[cfg(feature = "validate")]
+use super::{TypeError, ValidationError};
+
+#[cfg(feature = "validate")]
+use std::{convert::TryInto, hash::Hash, num::NonZeroU32};
+
+#[cfg(feature = "validate")]
+impl super::Validator {
+    /// Validates that all handles within `module` are:
+    ///
+    /// * Valid, in the sense that they contain indices within each arena structure inside the
+    /// [`crate::Module`] type.
+    /// * No arena contents contain any items that have forward dependencies; that is, the value
+    ///     associated with a handle only may contain references to handles in the same arena that
+    ///     were constructed before it.
+    ///
+    /// By validating the above conditions, we free up subsequent logic to assume that handle
+    /// accesses are infallible.
+    ///
+    /// # Errors
+    ///
+    /// Errors returned by this method are intentionally sparse, for simplicity of implementation.
+    /// It is expected that only buggy frontends or fuzzers should ever emit IR that fails this
+    /// validation pass.
+    pub(super) fn validate_module_handles(module: &crate::Module) -> Result<(), ValidationError> {
+        let &crate::Module {
+            ref constants,
+            ref entry_points,
+            ref functions,
+            ref global_variables,
+            ref types,
+        } = module;
+
+        // NOTE: Types being first is important. All other forms of validation depend on this.
+        for (this_handle, ty) in types.iter() {
+            let &crate::Type {
+                ref name,
+                ref inner,
+            } = ty;
+
+            let validate_array_size = |size| {
+                match size {
+                    crate::ArraySize::Constant(constant) => {
+                        let &crate::Constant {
+                            name: _,
+                            specialization: _,
+                            ref inner,
+                        } = constants.try_get(constant)?;
+                        if !matches!(inner, &crate::ConstantInner::Scalar { .. }) {
+                            return Err(ValidationError::Type {
+                                handle: this_handle,
+                                name: name.clone().unwrap_or_default(),
+                                source: TypeError::InvalidArraySizeConstant(constant),
+                            });
+                        }
+                    }
+                    crate::ArraySize::Dynamic => (),
+                };
+                Ok(this_handle)
+            };
+
+            match inner {
```

Comment:
It's Naga style to match `*inner`, drop the references from all the patterns, and use `ref` as appropriate to avoid moves. This is Clippy lint [match_ref_pats](https://rust-lang.github.io/rust-clippy/master/index.html#match_ref_pats), so I don't know why Clippy isn't complaining about this.

Reply:
If `clippy` doesn't or can't enforce this style…is it worth enforcing manually? 🤔

Changed as suggested with af211507.